### ISSUE TITLE
More DynamoDB support

### ISSUE THREAD

**Issue by mattroberts297** [aws/aws-sam-cli#219]
Great work - this is really useful - thanks! I have a feature request. It would be nice if `aws-sam-local` could take care of creating DynamoDB tables, routing traffic to them, and managing event source mappings between DynamoDB streams and lambdas locally e.g.

```
// All running locally:
API Gateway -> Front End Lambda -> DynamoDB -> DynamoDB Stream -> Back End Lambda
```

I think this would be quite involved, but I think it should be possible e.g. spin up a docker container with `DynamoDB` running, intercept calls from the `Front End Lambda` destined for AWS and send them to the container with `DynamoDB` instead, have a process consuming from the `DynamoDB Stream` and publishing to the `Back End Lambda` per the template.

If there is interest then I can create a simple template that works in AWS demonstrating the above so that we can discuss the feasibility /  mechanics of the change in more detail. Sorry in advance if this has already been proposed / discussed (I searched, but couldn't find anything).

*Edit: Thanks for the prioritisation and up-votes.*

**Comment by thoeni** [comment]
Hi there, I came across this issue while I was searching for dynamodb support as I started working on it and now I have aws-sam-local spinning up a dynamodb docker container, and killing it when SIGTERM...

Just started looking into lambdas, so if you wanna have a chat I’m more than happy to...

**Comment by Nr18** [comment]
From a discussion on the AWS Dev Slack @heitorlessa shared the following with me when i questioned this:

    $ docker network create sam-demo
    $ docker run -d -v "$PWD":/dynamodb_local_db -p 8000:8000 --network sam-demo --name dynamodb cnadiminti/dynamodb-local   

<<<bootstrap dynamodb table>>>

    $ sam local start-api --docker-network sam-demo

<<<change your code to use endpoint_url="http://dynamodb:8000">>>

This does allow you to use DynamoDB locally but it requires some additional steps would be great if SAM Local could do the same for DynamoDB as it does for the API Gateway.

The DynamoDB info is in the table so it could just scan the template and if there are dynamodb resources spin up the docker container with dynamodb local and create a table for each dynamodb table in the template.

Just wondering if you would use `dynamodb.us-east-1.amazonaws.com` as a name and:
- run over HTTPS
- include a self signed SSL certificate (will boto or any other sdk allow a self signed certificate or has an option to do so?)

That would remove the last part of changing the endpoint.

**Comment by abbottdev** [comment]
Can sam cli map endpoint-urls to the underlying s3 commands? e.g. I was wondering whether, until sam cli is feature complete, whether a bridge to allow interop with localstack would help here?

For example, is it possible to set the --endpoint-url for different aws services using sam cli? I know you can set it if you're using aws cloudformation, aws s3, etc. directly 

If I'm able to append `--s3-endpoint-url` or similar that gets mapped into an `--endpoint-url` on the corresponding aws cli commands - this would allow people to use and opt-in/out of `sam cli` services until there's more feature parity?

**Comment by fiddur** [comment]
> This does allow you to use DynamoDB locally

Being able to access DynamoDB is great, thanks for the instructions.

However, it doesn't give us the events, so a lambda triggered by dynamo event stream still won't trigger...  and other than making a separate service frequent polling the dynamodb, I don't see how to set that up locally.

**Comment by devrelm** [comment]
Related (closed) issue: #105

**Comment by sriram-mv** [comment]
There are great workarounds posted on this issue.
There are no plans to support a local version of DynamoDB natively in our tooling at this point in time.
